docs(ExampleContracts): add an LN contract (#14)
ivy-lang/ivy-bitcoin: docs/language/ExampleContracts.md

What happens if Bob disappears, or refuses to sign any transactions? That's what the `timeout` clause is for. After the channel's expiration time, Alice can call the `timeout` clause to reclaim all of the Bitcoin she used to prefund the channel. This prevents Alice's money from being locked up forever if Bob refuses to cooperate.

> ## LightningContractForPunishmentPaymentChannel

Reviewer: ```suggestion
## ToLocal
```

> This contract can be used for creating Lightning-Network payment channels. It is similar to the previous one because it includes a timeout clause (the last one) with a single signature, however, this timeout is not for the depositor of the funds to claim them in case the other party is not cooperative, but it's a time-window left open for any disputes to be presented after one party tries to close the channel. The dispute could originate from the fact that the party that sent the on-chain transaction to close the channel could be trying to propgrate a state of the channel which is not valid anymore (as it might have been invalidated by subsequent off-chain transactions); in such a case the party that didn't attempt to close the channel has this timeout available to proof this and make a punishment transaction effective, that would make her receive all the funds in the channel via providing the preimage of the hash that was used to initially create this contract (the first clause). More info: https://blog.chainside.net/understanding-payment-channels-4ab018be79d4#2435

Reviewer: ```suggestion
This contract can be used as part of a bidirectional payment channel similar to those used in the Lightning Network, as described [here](https://github.com/lightningnetwork/lightning-rfc/blob/master/03-transactions.md#to_local-output). This contract would be used as an output of a commitment transaction (which represents a state that the parties agree to off-chain).

It is similar to the previous contract because it includes a timeout clause (the last one) with a single signature. However, this timeout is not for the depositor of the funds to claim them in case the other party is not cooperative, but is a time window left open for any disputes to be presented after one party tries to close the channel. Such a dispute can arise if the party that sent the on-chain transaction to close the channel is trying to propagate a state of the channel which is not valid anymore (as it might have been invalidated by subsequent off-chain transactions). In such a case, the party that didn't attempt to close the channel has this timeout available to withdraw all the funds in a punishment transaction, using the revocation private key that she learned when the channel state was invalidated.
```

```

> contract LightningContractForPunishmentPaymentChannel(

Reviewer: ```suggestion
contract ToLocal(
```

>   channelCloserParty: PublicKey,

Reviewer: ```suggestion
  localDelayedPubKey: PublicKey,
```

>   potentialPunisherParty: PublicKey,

Reviewer: ```suggestion
```

>   channelCloserSecret: Sha256(Bytes),

Reviewer: ```suggestion
  revocationPubKey: PublicKey,
```

>   revocationTimeWindow: Duration,

Reviewer: ```suggestion
  toSelfDelay: Duration,
```

  val: Value
) {

>   clause reveal(potentialPunisherPartySig: Signature, string: Bytes) {

Reviewer: ```suggestion
  clause reveal(revocationSig: Signature) {
```

>     verify checkSig(potentialPunisherParty, potentialPunisherPartySig)

Reviewer: ```suggestion
    verify checkSig(revocationPubKey, revocationSig)
```

>     verify sha256(string) == channelCloserSecret

Reviewer: ```suggestion
```

    unlock val
  }

>   clause timeout(channelCloserPartySig: Signature) {

Reviewer: ```suggestion
  clause timeout(localDelayedSig: Signature) {
```

>     verify checkSig(channelCloserParty, channelCloserPartySig)

Reviewer: ```suggestion
    verify checkSig(localDelayedPubKey, localDelayedSig)
```

>     verify older(revocationTimeWindow)

Reviewer: ```suggestion
    verify older(toSelfDelay)
```
